mlx-nn/src/losses.rs

```diff
@@ -0,0 +1,1281 @@
+//! Loss functions
+
+use mlx_macros::option_builder;
+use mlx_rs::{
+    array,
+    error::Exception,
+    ops::{
+        abs, clip, exp, indexing::take_along_axis, log, log_add_exp, log_sum_exp, maximum, minimum,
+        multiply, power, r#where, sqrt, square, sum,
+    },
+    Array,
+};
+
+#[inline]
+fn check_shape(
+    left: &Array,
+    right: &Array,
+    left_ident: &str,
+    right_ident: &str,
+) -> Result<(), Exception> {
+    if left.shape() != right.shape() {
+        return Err(Exception::custom(format!(
+            "The shape of the {} ({:?}) does not match the shape of the {} ({:?})",
+            left_ident,
+            left.shape(),
+            right_ident,
+            right.shape()
+        )));
+    }
+    Ok(())
+}
+
+/// Different types of loss reductions
+#[derive(Debug, Clone, Copy)]
+pub enum LossReduction {
+    /// No reduction is applied.
+    None,
+    /// The sum of the output will be computed.
+    Sum,
+    /// The mean of the output will be computed.
+    Mean,
+}
+
+impl LossReduction {
+    /// Reduces the loss according to the reduction type.
+    pub fn reduce(&self, loss: Array) -> Result<Array, Exception> {
+        match self {
+            LossReduction::None => Ok(loss),
+            LossReduction::Sum => Ok(loss.sum(None, None)?),
+            LossReduction::Mean => Ok(loss.mean(None, None)?),
+        }
+    }
+}
+
+/// Optional parameters for the `cross_entropy` function.
+#[derive(Debug, Clone, Default)]
+#[option_builder]
+pub struct CrossEntropyOptions<'a> {
```

Comment:
Tried a different way to handle optional args. Took this approach recommended in azure sdk guideline for rust. Not too sure if this is an overkill.

@dcvz WDYT

Reply:
It's interesting for sure.. it's not too bad -- remind's a bit of Burn's builders. Have you seen those? The syntax there IMO is slightly better, but I think the ideas are very similar.

Here's for example a declaration: https://github.com/oxideai/diffusers-burn/blob/d52deb45110f721d8e8cb347d0e56ef80c56fc28/src/models/attention.rs#L94

and the usage here: https://github.com/oxideai/diffusers-burn/blob/d52deb45110f721d8e8cb347d0e56ef80c56fc28/src/models/attention.rs#L257

WDYT?

Reply:
Interesting. Their `CrossEntropy` is a struct and the actual computation takes place in `fn forward`. We probably don't want to call it `forward` just to avoid confusion with the `Module` trait if we were to adapt their approach?